mule-artifact-it/mule-packager-it/src/test/java/integration/test/mojo/AdditionalPluginDependenciesProcessSourcesMojoTest.java

```diff
@@ -117,6 +119,56 @@ public void muleAppWithMultiplePluginsWithSameAdditionalDependenciesWithDifferen
     assertArtifactInstalled(appLocation, GENERATED__DEPENDENCY_Z_V2_PATH_SUFFIX);
   }
 
+  /**
+   * Validates that if a mule-plugin defined additional libraries those are taken into account.
+   */
+  @Test
+  public void muleAppWithReusableAppPlugin() throws Exception {
+    final String appName = "mule-app-plugin-with-reusable-app";
+    final String appLocation = getAppLocation(appName);
+    processSourcesOnProject(appLocation);
+    List<String> generatedAppClassLoaderModelFileContent = getFileContent(getCorrectGeneratedClassloaderModelPath(appLocation));
+    List<String> expectedAppClassLoaderModelFileContent =
+        getFileContent(getCorrectExpectedClassloaderModelPath(appLocation, appName));
+    assertThat(generatedAppClassLoaderModelFileContent, equalTo(expectedAppClassLoaderModelFileContent));
+    assertArtifactInstalled(appLocation, GENERATED__DEPENDENCY_X_PATH_SUFFIX);
+    assertArtifactInstalled(appLocation, GENERATED__DEPENDENCY_Z_PATH_SUFFIX);
+  }
+
+  /**
+   * Validates that additional libraries defined at the mule application level overrides all other additional libraries from mule
+   * plugins.
+   */
+  @Test
+  public void muleAppWithReusableAppPluginAndOverridedAdditionalDependency() throws Exception {
+    final String appName = "mule-app-plugin-with-reusable-app-and-additional-dep-x-v2";
+    final String appLocation = getAppLocation(appName);
+    processSourcesOnProject(appLocation);
+    List<String> generatedAppClassLoaderModelFileContent = getFileContent(getCorrectGeneratedClassloaderModelPath(appLocation));
+    List<String> expectedAppClassLoaderModelFileContent =
+        getFileContent(getCorrectExpectedClassloaderModelPath(appLocation, appName));
+    assertThat(generatedAppClassLoaderModelFileContent, equalTo(expectedAppClassLoaderModelFileContent));
+    assertArtifactInstalled(appLocation, GENERATED__DEPENDENCY_X_V2_PATH_SUFFIX);
+  }
+
+  /**
+   * Validates that if two mule-plugin have the same additional library then the latest one is going to be used.
+   */
+  @Test
+  public void muleAppWithReusableAppAndUsingAnotherReusableAppWithANewestXVersion() throws Exception {
```

Comment:
Reusable apps are mule plugins and follow exactly the same rules as mule-plugins. The reusable application that adds the dependency to a mule-plugin can also add additionalDependencies for that plugin so when you reuse it from another application you don't have to set those additionalDependencies manually.

What this test case is validating is that if app A dependes on reusable app B and C, and B has an additional dependency to X 1.0 and C has an additional dependency to X 2.0 then X 2.0 gets used.

Reply:
They behave as mule-plugins but they are not isolated as mule-plugins, right? 

For what you answered in the other PR, i understand that by having A,B and C, it's like having a virtual D app with all plugins from A, B and C as dependencies. 

There is one more case that i think should be tested though. 

Consider app A has reusable apps B and C. Both of them depend on the same plugin X. B depends on X v1 and C depends on X v2. 
If i understood correctly , Xv2 will remain because it has the bigger version.
Now consider that B and C declare an additionalPluginDependency Y for their X dependency. 
B declares Y v2 and C declares Y v1. 
Which version should remain? 
Yv2 is bigger than Yv1 but it's added to Xv1 that is smaller than Xv2. 

Don't know if this is actually and issue but it may be worth thinking about it.